File: client/typescript/src/workspace/branch.ts
import type { Revision } from "../_generated/repository/revision.js";
import type { FileSelection } from "../_generated/workspace/branch.js";
import type { EditBranchRequest } from "../_generated/workspace/service/source.js";
import type { Workspace } from "./workspace.js";

/** One named workspace revision. */
export class Branch {
    /** Workspace containing this branch. */
    readonly workspace: Workspace;
    /** Branch name. */
    readonly name: string;

    /** Bind one workspace branch. */
    constructor(workspace: Workspace, name: string) {
        this.workspace = workspace;
        this.name = name;
    }

    /** Read this branch's current revision. */
    async revision(): Promise<Revision> {
        const response = await this.workspace.client.branchRevision({
            root: this.workspace.root,
            name: this.name,
        });

        return response.value;
    }

    /** Commit source edits to one exact branch revision. */
    async edit(request: Omit<EditBranchRequest, "root" | "name">) {
        const response = await this.workspace.client.editBranch({
            root: this.workspace.root,
            name: this.name,
            ...request,
        });

        return response.value;
    }

    /** Save selected files to physical workspace state. */
    async save(files: FileSelection) {
        const [revision, physical] = await Promise.all([
            this.revision(),
            this.workspace.revision(),
        ]);

        const response = await this.workspace.client.saveBranch({
            root: this.workspace.root,
            name: this.name,
            revision,
            physical,
            files,
        });

        return response.value;
    }

    /** Restore selected files from physical workspace state. */
    async restore(files: FileSelection) {
        const [revision, physical] = await Promise.all([
            this.revision(),
            this.workspace.revision(),
        ]);

        const response = await this.workspace.client.restoreBranch({
            root: this.workspace.root,
            name: this.name,
            revision,
            physical,
            files,
        });

        return response.value;
    }

    /** Watch this branch for exact committed transitions. */
    watch() {
        return this.workspace.client.watchBranch({
            root: this.workspace.root,
            name: this.name,
        });
    }

    /** Remove this branch. */
    async remove(): Promise<void> {
        await this.workspace.client.removeBranch({
            root: this.workspace.root,
            name: this.name,
        });
    }
}

File: language/workspace/src/workspace/branch.rs
use std::path::PathBuf;

use serde::{Deserialize, Serialize};
use tspp_repository::Revision;
use tspp_serde::Reflect;

use crate::{Error, Watch, Workspace};

/// One named workspace revision.
#[derive(Debug, Clone, PartialEq, Eq, Serialize, Deserialize, Reflect)]
pub struct Branch {
    /// Branch name.
    pub name: String,
    /// Current branch revision.
    pub revision: Revision,
}

/// Files selected by one branch operation.
#[derive(Debug, Clone, PartialEq, Eq, Serialize, Deserialize, Reflect)]
pub enum FileSelection {
    /// Every changed file.
    All,
    /// Selected paths within the workspace.
    Paths(Vec<PathBuf>),
}

impl Workspace {
    /// List every branch in name order.
    pub fn branches(&self) -> Result<Vec<Branch>, Error> {
        let state = self.lock()?;
        let mut branches = state
            .branches
            .iter()
            .map(|(name, revision)| Branch {
                name: name.clone(),
                revision: revision.revision(),
            })
            .collect::<Vec<_>>();
        branches.sort_unstable_by(|left, right| left.name.cmp(&right.name));

        Ok(branches)
    }

    /// Create one branch at an existing revision.
    pub fn create_branch(&self, name: String, revision: Revision) -> Result<Branch, Error> {
        let mut state = self.lock()?;
        if state.branches.contains_key(&name) {
            return Err(Error::BranchExists { name });
        }

        let pin = self.repository.pin(revision)?;
        let branch = Branch {
            name: name.clone(),
            revision,
        };
        state.branches.insert(name, pin);

        Ok(branch)
    }

    /// Return one branch's current revision.
    pub fn branch_revision(&self, name: &str) -> Result<Revision, Error> {
        let state = self.lock()?;
        let branch = state.branch(name)?;

        Ok(branch.revision())
    }

    /// Remove one branch.
    pub fn remove_branch(&self, name: &str) -> Result<(), Error> {
        let mut state = self.lock()?;
        state
            .branches
            .remove(name)
            .ok_or_else(|| Error::MissingBranch {
                name: name.to_string(),
            })?;
        self.watch.lock().remove_branch(name);

        Ok(())
    }

    /// Watch one workspace branch.
    pub fn watch_branch(&self, name: &str) -> Result<Watch, Error> {
        let state = self.lock()?;
        let revision = state.branch(name)?.revision();

        Watch::new(
            self.root.clone(),
            Some(name.to_string()),
            revision,
            &self.repository,
            self.watch.clone(),
        )
    }
}
